src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+            // (which can happen if the user deletes down to a single character and we
+            // include everything), then we just soft select the first item.
+            if (bestFilterResult != null)
+            {
+                // Only hard select this result if it's a prefix match
+                // We need to do this so that
+                // * deleting and retyping a dot in a member access does not change the
+                //   text that originally appeared before the dot
+                // * deleting through a word from the end keeps that word selected
+                // This also preserves the behavior the VB had through Dev12.
+                hardSelect = bestFilterResult.Value.VSCompletionItem.FilterText.StartsWith(filterText, StringComparison.CurrentCultureIgnoreCase);
+                index = filterResults.IndexOf(bestFilterResult.Value);
+            }
+            else
+            {
+                index = 0;
+                hardSelect = false;
+            }
+
+            return new AsyncCompletionData.FilteredCompletionModel(
+                highlightedList, index, filters,
+                hardSelect ? AsyncCompletionData.UpdateSelectionHint.Selected : AsyncCompletionData.UpdateSelectionHint.SoftSelected,
+                centerSelection: true, uniqueItem: null);
+        }
+
+        private AsyncCompletionData.FilteredCompletionModel HandleAllItemsFilteredOut(
+            AsyncCompletionData.CompletionTriggerReason triggerReason,
+            ImmutableArray<AsyncCompletionData.CompletionFilterWithState> filters,
+            bool hasActiveFilters,
+            CompletionRules completionRules)
+        {
+            AsyncCompletionData.UpdateSelectionHint selection;
+            if (triggerReason == AsyncCompletionData.CompletionTriggerReason.Insertion)
+            {
+                // If the user was just typing, and the list went to empty *and* this is a 
+                // language that wants to dismiss on empty, then just return a null model
+                // to stop the completion session.
+                if (completionRules.DismissIfEmpty)
+                {
+                    return null;
+                }
+
+                selection = AsyncCompletionData.UpdateSelectionHint.NoChange;
+            }
+            else
+            {
+                // If the user has turned on some filtering states, and we filtered down to
+                // nothing, then we do want the UI to show that to them.  That way the user
+                // can turn off filters they don't want and get the right set of items.
+
+                // If we are going to filter everything out, then just preserve the existing
+                // model (and all the previously filtered items), but switch over to soft
+                // selection.
+                selection = hasActiveFilters ? AsyncCompletionData.UpdateSelectionHint.NoChange : AsyncCompletionData.UpdateSelectionHint.SoftSelected;
```

Comment:
this doesn't seem to match old code. in the old code we always set things to soft selection in the corresponding method.  is the change here intentional?